Initial implementation (#1)
projectsyn/component-lieutenant: docs/modules/ROOT/pages/references/parameters.adoc

[horizontal]
type:: string
default:: `lieutenant`

The namespace in which to deploy this component.

== `images`

[horizontal]
type:: dictionary

Dictionary containing the container images used by this component.

== `operator.common_labels`

[horizontal]
type:: dict
default: ::
[source,yaml]
----
app.kubernetes.io/name: 'lieutenant-operator'
app.kubernetes.io/part-of: 'project-syn'
----

Common labels to add to each resource of the Lieutenant Operator.
Will be used as a label selector for pods.

== `operator.default_deletion_policy`

[horizontal]
type:: string
default:: `Archive`

Sets what deletion policy for external resources (Git, Vault) should be used by default.
One of `Archive`, `Delete`, `Retain`.

== `operator.deletion_protection`

[horizontal]
type:: bool
default:: `false`

Defines whether the annotation to protect for accidental deletion should be set by default.

== `operator.default_global_git_repo`

[horizontal]
type:: string
default:: ``

> URL of the default global configuration git repository.

Reviewer: ```suggestion
URL of the default global configuration Git repository.
```

> Its value will be applied to `tenant.Spec.GlobalGitRepoURL`.

Reviewer: IIRC how this works:

```suggestion
The value of this parameter will be used as the default value for `.spec.globalGitRepoURL` on new tenant objects.
```

== `operator.vault.enabled`

[horizontal]
type:: bool
default:: `true`

Whether to enable the Lieutenant Operator Vault integration.
If set to true, `operator.vault.addr` and `operator.vault.path` must be set to point to a valid Vault instance and a valid KV secret engine of that Vault instance.

== `operator.vault.addr`

[horizontal]
type:: string
default:: `vault.todo`

Address to the Vault instance.

== `operator.vault.path`

[horizontal]
type:: string
default:: `kv`

The mount path of the KV secret engine to be used.

== `api.common_labels`

[horizontal]
type:: dict
default: ::
[source,yaml]
----
app.kubernetes.io/name: 'lieutenant-api'
app.kubernetes.io/part-of: 'project-syn'
----

> Common label to add to each resource of the Lieutenant API.

Reviewer: ```suggestion
Common labels to add to each resource of the Lieutenant API.
```

Will be used as a label selector for pods.

== `api.default_githost`

[horizontal]
type:: string
default:: ``

The githost to be used by default for new tenants.

== `api.ingress.host`

[horizontal]
type:: string
default:: `lieutenant.todo`

Defines the FQDN of the API ingress, should be overwritten on the cluster level.

== `api.ingress.annotations`

[horizontal]
type:: dict
default:: {}

The annotations added to the created ingress
Needs to be set according to the deployed ingress controller.

[source,yaml]
----
annotations:
  cert-manager.io/cluster-issuer: letsencrypt-production
  kubernetes.io/ingress.class: nginx
  nginx.ingress.kubernetes.io/cors-allow-credentials: 'true'
  nginx.ingress.kubernetes.io/cors-allow-methods: GET, POST, DELETE
  nginx.ingress.kubernetes.io/cors-allow-origin: http://localhost:8080
  nginx.ingress.kubernetes.io/enable-cors: 'true'
----

== `api.ingress.tls`

[horizontal]
type:: bool
default:: `true`

> Whether to enable tls for this ingress.

Reviewer: ```suggestion
Whether to enable TLS for the ingress.
```

This requires either to set the correct `cert-manager` annotations or to add the certificate manually to the secret `lieutenant-api-cert`.

== `api.users`

[horizontal]
type:: list
default: ::

[source,yaml]
----
users:
  - kind: ServiceAccount
    name: lieutenant-api-user
----

A list of users that have permission to access the API

> These entries translate to kubernetes subjects and can reference a `Group`, `User`, or `ServiceAccount`.

Reviewer: ```suggestion
These entries translate to Kubernetes subjects and can reference a `Group`, `User`, or `ServiceAccount`.
```

> Each specified `ServiceAccount` will be created.

Reviewer: ```suggestion
For entries with type `ServiceAccount` the component will create the corresponding service account object.
```

== `tenant_rbac`

[horizontal]
type:: dict
default:: {}

Role based access control to the created tenant resources.
Lieutenant creates a `Role` for each tenant.

[source,yaml]
----
tenant_rbac:
  t-foo-324
    - name: 'sa-bar'
      kind: 'ServiceAccount'
    - name: 'u-bar-1'
      kind: 'User'
  t-foo-1
    - name: 'g-buzz'
      kind: 'Group'
    - name: 'u-bar-1'
      kind: 'User'
----

The example configuration above will grant user `u-bar-1` and service account `sa-bar` read access to all Clusters owned by Tenant `t-foo-324`.
And it will grant group `g-buzz` and user `u-bar-1` read access to all Clusters owned by Tenant `t-foo-1`.

This can usually only be configured after the initial setup of Lieutenant.

== `githosts`

[horizontal]
type:: dict
default:: {}

A list of GitLab instances Lieutenant will be able to connect to for repository creation.

> A GitLab toke with `api` scope need to be accessible through Vault.

Reviewer: ```suggestion
A GitLab token with `api` scope need to be accessible through Vault.
```

default:: `docker.io/projectsyn/lieutenant-api:v0.6.0`

The lieutenant-api container image to deploy.

== `api.common_labels`

[horizontal]
type:: dict
default: ::
[source,yaml]
----
app.kubernetes.io/name: 'lieutenant-api'
app.kubernetes.io/part-of: 'project-syn'
----

Common label to add to each resource of the lieutenant-api. Will be used
as a label selector for pods.

== `api.default_githost`

[horizontal]
type:: string
default:: ``

The githost to be used by default for new tenants.

== `api.steward_image`

[horizontal]
type:: string
default:: `docker.io/projectsyn/steward:v0.3.1`

Image to use in generated `Steward` deployment manifests.

== `api.ingress.host`

[horizontal]
type:: string
default:: `lieutenant.todo`

Defines the FQDN of the API ingress, should be overwritten on the cluster level.

== `api.ingress.annotations`

[horizontal]
type:: dict
default:: {}

The annotations added to the created ingress. Needs to be set according to the
deployed ingress controller.

[source,yaml]
----
annotations:
  cert-manager.io/cluster-issuer: letsencrypt-production
  kubernetes.io/ingress.class: nginx
  nginx.ingress.kubernetes.io/cors-allow-credentials: 'true'

>   nginx.ingress.kubernetes.io/cors-allow-methods: GET, POST, DELETE

Reviewer: Definitely not sufficient for the API, at least `PATCH` needs to be allowed too.

Author: I mean you're right, the API definitely needs `PATCH`, but this config currently runs in [production](https://git.vshn.net/syn/synfra-apps/-/blob/master/lieutenant-prod/ingress.yaml).

Reviewer: well, then it's fine I guess, I didn't check the production config

  nginx.ingress.kubernetes.io/enable-cors: 'true'
----

== `api.ingress.users`

[horizontal]
type:: list
default: ::

[source,yaml]
----
users:
  - kind: ServiceAccount
    name: api-access-vshn-portal-prod
----

A list of users that have permission to access the API. These entries translate
to kubernetes subjects and can reference a `Group`, `User`, or `ServiceAccount`.
Each specified `ServiceAccount` will be created.

== `api.tenant_rbac`

[horizontal]
type:: list
default:: []

Role based access control to the created tenant resources. Lieutenant creates a `Role`
for each tenant.

[source,yaml]
----
tenant_rbac:
  - tenant: 't-foo-324'
    subjects:
      - name: 'sa-bar'
        kind: 'ServiceAccount'
      - name: 'u-bar-1'
        kind: 'User'
  - tenant: 't-foo-1'
    subjects:
      - name: 'g-buzz'
        kind: 'Group'
      - name: 'u-bar-1'
        kind: 'User'
----

The example configuration above will grant user `u-bar-1` and service account `sa-bar` read access to
all Clusters owned by Tenant `t-foo-324`. And it will grant group `g-buzz` and user `u-bar-1`
read access to all Clusters owned by Tenant `t-foo-1`.

> This can usually only be configured after the initial setup of lieutenant.

Reviewer: ```suggestion
This can usually only be configured after the initial setup of Lieutenant.
```

Please make sure you capitalize Lieutenant consistently in the docs.

== `api.githosts`

[horizontal]
type:: list
default:: []

A list of GitLab instances Lieutenant will be able to connect to for repository creation.

> A GitLab toke with `api` scope need to be accessible through vault. See the

Reviewer: ```suggestion
A GitLab toke with `api` scope need to be accessible through Vault. See the
```

Please check capitalization for Vault in the docs
